### ISSUE TITLE
charon-nm: only a single CA cert file is loaded from "server certificate" file

### ISSUE THREAD

**Issue by LuKePicci** [strongswan/strongswan#2219]
#### Root Issue
When a certificate file is selected for the server, the nm backend loads only a single cert from that file, even if it is a PEM bundle of both CA certs.

#### Affected use cases
This breaks PEAP/EAP-TLS/EAP-TTLS/ecc. when the IPsec gateway authenticates itself via a different CA from the one behind the TLS handshake within EAP, which is quite common when using a RADIUS server.

#### Code affected
https://github.com/strongswan/strongswan/blob/f8e6fd30de25fa350f55bf1dc1e0c02b1c50460b/src/charon-nm/nm/nm_service.c#L769-L789

#### Workaround
When no cert file is selected, the nm backend loads all certs from the system CA folder. Putting the same CA certs bundle PEM file in that folder makes TLS-based EAP methods working again.

#### Expected fix
All certs from inside that file should be loaded as trusted CAs or gateway/server cert.

**Comment by harridu** [comment]
https://bugs.debian.org/853266 doesn't provide sufficient details, but it might be related.

Maybe its oversimplified, but I find it difficult to understand that a directory providing 2 certificates is supported, while a file providing the same 2 certificates is not. Just my $0.02.

**Comment by tobiasbrunner** [comment]
> https://bugs.debian.org/853266 doesn't provide sufficient details, but it might be related.

Don't think so. Not only was that created long before basic support for TLS-based EAP methods was added to charon-nm with bc3eda99bac0e18f13316c685d720f4148294364 in 2020, the main issue is that [`libcharon-extra-plugins` does not ship the _eap-peap_ plugin](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=843180) (it does ship the _eap-ttls_ plugin, though).



> Maybe its oversimplified, but I find it difficult to understand that a directory providing 2 certificates is supported, while a file providing the same 2 certificates is not.

strongSwan's certificate parsers can only handle a single certificate per file. So loading a directory with multiple files, each containing a trusted certificate, is straight-forward, loading multiple certificates from a single file is not.

**Comment by minfrin** [comment]
Found this issue because I am currently trying to automate certificate configs, and automation is currently very difficult.

To make strongswan work, it appears you need to calculate your chain of certs from leaf to root (easy), place them in separate files, leaf, intermediate1, intermediate2, etc, root (not easy, you have n intermediates, therefore n files, therefore a lot of complexity that should ideally be avoided), point the config at each of these files individually (not easy).

At the same time, ask any LLM for help and you get confident responses about multiple certs being supported, because the average answer across most services is multiple certs per file.

The config directories hint that pkcs12 is supported, but there is no docs I can find that describe how to reference such a file:

```
Comma-separated list of certificate candidates to use for authentication. The certificates may use a relative path from the [swanctl/x509](https://docs.strongswan.org/docs/latest/swanctl/swanctlDir.html) directory or an absolute path. The certificate used for authentication is selected based on the received certificate request payloads. If no appropriate CA can be located, the first certificate is used
```

**Comment by tobiasbrunner** [comment]
@minfrin Are you actually using the NM plugin? If so, you generally only need to configure a single root CA certificate (this issue is about scenarios where you might need another when using TLS-based EAP methods). Clients should receive the required intermediate CA certificates from the server. Or are you configuring the charon(-systemd) daemon with swanctl? If so, that's off-topic here and you might want to start a [discussion thread](https://github.com/strongswan/strongswan/discussions).

**Comment by minfrin** [comment]
Not the NM plugin no, this is native config files under /etc/strongswan/swanctl.

Automate means "certificate lifetimes are getting shorter, this can no longer be done by hand any more". "By hand" means chopping up certs and intermediates and constructing chains by hand. I built redwax-tool to handle all the different ways different daemons want their certs, and am happy to extend redwax-tool to accommodate what strongswan needs, but I need to understand strongswan first.

It seems there are levels to things. At the base level, if pem files / pkcs12 files exist under swanctl they are hoovered up and are available to form trust chains when called for.

At a higher level, it's hard to isolate each certificate in a connection, so it is unambiguous which one you use. You can refer to a pem file by name (good), but not a pkcs12 file (could not make that work). You can refer to certs by their subject, but two different certs with different trust chains could have the same subject (think last year's cert came from Digitcert but this year's is Lets Encrypt).

To automate this I need to say "use this cert and this chain and no other chain" in the local.certs field. The local.cacerts field confuses me, it appears to be needed so that strongswan server trusts itself before strongswan is willing to use the cert in local.certs.

**Comment by tobiasbrunner** [comment]
> You can refer to a pem file by name (good), but not a pkcs12 file (could not make that work).

Correct, PKCS#12 containers are primarily supported as key containers (certificates are extracted and sent to the daemon, but you can't directly reference them in a connection).

> You can refer to certs by their subject, but two different certs with different trust chains could have the same subject (think last year's cert came from Digitcert but this year's is Lets Encrypt).

That also makes it more flexible as you'll use the latest certificate without having to reload the config.

> To automate this I need to say "use this cert and this chain and no other chain" in the local.certs field. The local.cacerts field confuses me, it appears to be needed so that strongswan server trusts itself before strongswan is willing to use the cert in local.certs.

While `local.cacerts` technically exists, it's not really used for anything (it's not even documented).  `cacerts` is an option for `remote` sections, to enforce that client certificates are issued by a specific CA. The local chain doesn't have to be complete or even there to use the certificate, but you usually need to send intermediate CA certificates to peers (or validate their certificates). So you just have to make sure the certificates are loaded (e.g. by putting them or a symlink in `swanctl/x509ca`), then the local chain is automatically built starting from the local certificate (either hard-coded in the config or found via identity) based on key identifiers (or DNs as fallback).

**Comment by tobiasbrunner** [comment]
The `pem-bundles` branch adds support for loading PEM-encoded bundles as trust anchor in `charon-nm` (and via `swanctl` and a new `pki --bundle` utility). If anybody has time to try it, please do and report back. Thanks!
